src/app/request-handlers/spend-for-additional-points-request-handler.ts

```diff
@@ -0,0 +1,90 @@
+import { Inject, Injectable } from '@angular/core';
+import { RequestHandler } from './request-handlers';
+import type { SpendForAdditionalPointsTokenOption } from 'app/token-options/spend-for-additional-points-token-option';
+import type { SpendForAdditionalPointsRequest } from 'app/requests/spend-for-additional-points-request';
+import { CanvasService } from 'app/services/canvas.service';
+import { ProcessedRequest } from 'app/data/processed-request';
+import type { StudentRecord } from 'app/data/student-record';
+import type { TokenATMConfiguration } from 'app/data/token-atm-configuration';
+import { RequestHandlerGuardExecutor } from './guards/request-handler-guard-executor';
+import { addPercentOrPointsToCanvasGrade } from 'app/utils/canvas-grading';
+import { MultipleRequestsGuard } from './guards/multiple-requests-guard';
+import { ExcludeTokenOptionsGuard } from './guards/exclude-token-options-guard';
+import { SufficientTokenBalanceGuard } from './guards/sufficient-token-balance-guard';
+
+@Injectable()
+export class SpendForAdditionalPointsRequestHandler extends RequestHandler<
+    SpendForAdditionalPointsTokenOption,
+    SpendForAdditionalPointsRequest
+> {
+    constructor(@Inject(CanvasService) private canvasService: CanvasService) {
+        super();
+    }
+    public async handle(
+        configuration: TokenATMConfiguration,
+        studentRecord: StudentRecord,
+        request: SpendForAdditionalPointsRequest
+    ): Promise<ProcessedRequest> {
+        const { gradingType, pointsPossible } = await this.canvasService.getAssignmentGradingTypeAndPointsPossible(
+            configuration.course.id,
+            request.tokenOption.assignmentId
+        );
+        if (gradingType !== 'points' && gradingType !== 'percent') {
+            throw new Error(
+                `${request.tokenOption.assignmentName} doesn’t display its Grade as Points or as a Percent, and so its score cannot be added to.`
+            );
+        }
```

Comment:
I agree that checking for the grading type and publish state of assignment should be checked, but I believe it more belongs to the review functionality rather than checking it for every request. Currently for other token option types, the check is still done passively (i.e., if an API call returns an error, then somethings is wrong).

Reply:
But you're right that grading type check is necessary in this case given that posting a grade with invalid value will not get an error but instead make the grade become `null`. I think the responsibility to warn the instructor that something goes wrong should belong to the review functionality, but we definitely need a check for grading type before the API call to change the grade.